File: vademecum/code-src/sieciowe/TCP_-_serwer.py

import socket, select, signal, sys, os

MAX_CHILD = 5
QUERY_SIZE = 3
TIMEOUT = 13
BUF_SIZE = 4096

if len(sys.argv) != 2:
	print("USAGE: " + sys.argv[0] + " listenPort", file=sys.stderr)
	exit(1);

# obsługa sygnału o zakończeniu potomka
childNum = 0
def onChildEnd(s, f):
	print("odebrano sygnał o śmierci potomka")
	global childNum
	childNum -= 1
	os.waitpid(-1, os.WNOHANG);
signal.signal(signal.SIGCHLD, onChildEnd)

# utworzenie gniazd sieciowych ... SOCK_STREAM oznacza TCP
sfd_v4 = socket.socket(socket.AF_INET,  socket.SOCK_STREAM)
sfd_v6 = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)

# ustawienie opcji gniazda ... IPV6_V6ONLY=1 wyłącza korzystanie
# z tego samego socketu dla IPv4 i IPv6
sfd_v6.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 1)

# przypisanie adresów ...
# '0.0.0.0' oznacza dowolny adres IPv4 (czyli to samo co INADDR_ANY)
# '::' oznacza dowolny adres IPv6 (czyli to samo co in6addr_any)
sfd_v4.bind(('0.0.0.0', int(sys.argv[1])))
sfd_v6.bind(('::',      int(sys.argv[1])))

# określenie gniazd jako używanych do odbioru połączeń przychodzących
# (długość kolejki połączeń ustawiona na wartość QUERY_SIZE)
sfd_v4.listen(QUERY_SIZE)
sfd_v6.listen(QUERY_SIZE)

# funkcja zajmująca się odbieraniem połączeń i ich obsługą
def acceptConn(sfd):
	global childNum
	
	#  odebranie połączenia
	sfd_c, sAddr = sfd.accept()
	
	# weryfikacja ilości potomków
	if childNum >= MAX_CHILD:
		print("za dużo potomków - odrzucam połączenie od:", sAddr);
		sfd_c.send("Internal Server Error\r\n".encode())
		sfd_c.close()
		return
	
	# aby móc obsługiwać wiele połączeń rozgałęziamy proces
	pid = os.fork()
	if pid == 0:
		print("połączenie od:", sAddr)
		while True:
			# czekanie na dane z timeout'em
			# aby zabezpieczyć się przed atakiem DoS
			rd, _, _ = select.select([sfd_c], [], [], TIMEOUT)
			if sfd_c in rd:
				data = sfd_c.recv(BUF_SIZE)
				if not data:
					print("koniec połączenia od:", sAddr)
					break
				print("odebrano od", sAddr, ":", data.decode());
				sfd_c.send(data)
			else:
				print("timeout połączenia od:", sAddr)
				break
		# zamykanie połączenia
		sfd_c.shutdown(socket.SHUT_RDWR)
		sfd_c.close()
		sys.exit()
	else:
		childNum += 1

# czekanie na połączenia z użyciem select() w nieskończonej pętli
while True:
	sfd, _, _ = select.select([sfd_v4, sfd_v6], [], [])
	if sfd_v4 in sfd:
		acceptConn(sfd_v4)
	if sfd_v6 in sfd:
		acceptConn(sfd_v6)

File: vademecum/code-src/sieciowe/TCP_-_serwer.cpp

#include <sys/socket.h>
#include <netinet/ip.h>
#include <arpa/inet.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/wait.h>

#define BUF_SIZE 256
#define QUERY_SIZE 3
#define MAX_CHILD  5

int childNum = 0;
void onChildEnd(int sig) {
	puts("odebrano sygnał o śmierci potomka");
	--childNum;
	waitpid(-1, NULL, WNOHANG);
}

int main(int argc, char *argv[]) {
	int res;
	char buf[BUF_SIZE];
	
	if (argc != 2) {
		fprintf(stderr, "USAGE: %s listenPort\n", argv[0]);
		return EXIT_FAILURE;
	}
	
	// obsługa sygnału o zakończeniu potomka
	signal(SIGCHLD, &onChildEnd);
	
	// utworzenie gniazda sieciowego ... SOCK_STREAM oznacza TCP
	int sfd = socket(PF_INET6, SOCK_STREAM, 0);
	if (sfd<0) {
		perror("Error in socket()");
		return EXIT_FAILURE;
	}
	
	// ustawienie opcji gniazda ... IPV6_V6ONLY=0 umożliwia korzystanie z tego
	// samego socketu dla IPv4 i IPv6 (rozszerzenie Linux-owe - na niektórych
	// systemach nie jest możliwe dualne słuchanie przy pomocy socketu IPv6)
	int opt = 0;
    if (setsockopt(sfd, IPPROTO_IPV6, IPV6_V6ONLY, &opt, sizeof(opt)) < 0) {
		perror("setsockopt(IPV6_V6ONLY=0)");
		return EXIT_FAILURE;
	}
	
	// utworzenie struktury opisującej adres
	struct sockaddr_in6 serwer;
	serwer.sin6_family=AF_INET6;
	serwer.sin6_port=htons(atoi(argv[1]));
	// host to network short (porządek bajtów), są też dla long i network to host
	serwer.sin6_addr=in6addr_any;
	// in6addr_any oznacza że nasłuchujemy na każdym adresie IP danego hosta
	// (zamiast tego można określić konkretny adres)

	// przypisanie adresu ...
	if (bind(sfd, (struct sockaddr *) &serwer, sizeof(struct sockaddr_in6)) < 0) {
		perror("Error in bind()");
		return EXIT_FAILURE;
	}

	// określenie gniazda jako używanego do nasłuchiwania połączeń przychodzących
	if (listen(sfd, QUERY_SIZE) < 0) {
		perror("Error in listen()");
		return EXIT_FAILURE;
	}
	
	while(1) {
		// odebranie połączenia
		struct sockaddr_in6 from;
		socklen_t fromLen=sizeof(struct sockaddr_in6);
		int sfd2 = accept(sfd, (struct sockaddr *) &from, &fromLen);
		
		// weryfikacja ilości potomków
		if (childNum >= MAX_CHILD) {
			printf("za dużo potomków\n");
			// można tutaj także wysłać informacje o błędzie serwera do klienta
			close(sfd2);
			continue;
		}
		
		// aby móc obsługiwać wiele połączeń rozgałęziamy proces
		int pid = fork();
		if (pid == 0) {
			// proces potomny ... zajmuje się tylko nowym połączeniem,
			// a nie nasłuchaniem więc zamykamy sfd
			close(sfd);
			
			// konwersja adresu na postać napisową
			char fromStr[64];
			snprintf(fromStr, 64, "%s:%d",
				inet_ntop(AF_INET6, &(from.sin6_addr), buf, BUF_SIZE),
				ntohs(from.sin6_port)
			);
			
			// obsługa połączenia
			printf("połączenie od: %s\n", fromStr);
			FILE * net;
			net=fdopen(sfd2, "r+");
			// należałoby tutaj używać czekania z timeout'em ...
			// inaczej usługa jest podatna na ataki DoS
			while(fgets(buf, BUF_SIZE, net)) {
				printf("odebrano od %s: %s\n", fromStr, buf);
				fputs(buf, net);
			}
			fclose(net);
			printf("koniec połączenia od: %s\n", fromStr);
			
			// zakończenie połączenia i potomka
			close(sfd2);
			return 0;
		}
		
		if (pid == -1)
			perror("Error in fork()");
		
		++childNum;
		close(sfd2);
	}

	// zamkniecie gniazda
	close(sfd);
}
